Zigbee2mqtt service

cicoub13 opened the pull request:
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] If your changes modify the API (REST or Node.js), did you modify the API documentation? (Documentation is based on comments in code)
- [x] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/v4-website) and the [website](https://gladysassistant.com). https://github.com/GladysAssistant/v4-website/pull/63
- [x] Did you add fake requests data for the demo mode (`front/src/config/demo.json`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Zigbee2mqtt Service

Image to test is here: https://hub.docker.com/layers/cicoub13/gladys/dev-zigbee2mqtt/images/sha256-70311fba250f79e874eb7d6c06099a0d4a8aee3dc737fad2d839f1ae0a769256?context=explore

There are 2 commits related to the core of Gladys 
- Map USB productId and vendorId https://github.com/GladysAssistant/Gladys/pull/1098/commits/5e164f514d707890ab70e15f6806f9fe9a2acf8b
- Stop Docker Container in system https://github.com/GladysAssistant/Gladys/pull/1098/commits/5a1d757c368fd0c832a8532d265987e25b4d2ba4

Forum discussion: https://community.gladysassistant.com/t/zigbee2mqtt-image-docker-de-test-basee-gladys-v4/5746/245

R6n0 committed:
feat(zigbee2mqtt): Add service integration

R6n0 committed:
feat(core): Add Stop Docker container function in system

R6n0 committed:
feat(core): Add USB informations (productId, vectorId) for dongle differentiation

R6n0 committed:
feat(core): Exclude zigbee2mqtt-container.json from nodemon

cicoub13 committed:
fix(zigbee2mqtt): Remove USB disable_led and fix path

cicoub13 committed:
fix(zigbee2mqtt): Containers' names and detection

cicoub13 committed:
feat(zigbee2mqtt): Setup Page in Mobile

cicoub13 committed:
fix(zigbee2mqtt): MQTT clientId random suffix + mosquito 2

cicoub13 committed:
feat(zigbee2mqtt): Add TuYa, Lonsonho devices and fix models

cicoub13 committed:
feat(zigbee2mqtt): Device detection and not managed devices

cicoub13 committed:
test(zigbee2mqtt): Finish test :rocket:

cicoub13 committed:
fix(zigbee2mqtt): Feedbacks Louis

- Error message when scan returns no device
- Revamp feature and value convertion
- Ignore empty MQTT messages
- Put default name for each feature
- Add hectoPascal Unit
- Fix kWh unit

cicoub13 committed:
fix(zigbee2mqtt): Last fixes

- Manage websocket connection lost in Discover
- Remove devices without model from Discover
- Automatic feature name

LM1LC3N7 committed:
Update Xiaomi.js

Update:
* Aqara door & window contact sensor (MCCGQ11LM) to add temperature feature

Add:
* Aqara D1 single key wireless wall switch (WXKG06LM)

cicoub13 committed:
feat(zigbee2mqtt): Refactor and clean server code + doc

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/discover-page/DiscoverTab.jsx:
```diff
@@ -0,0 +1,82 @@
+            {props.zigbee2mqttDevices && (
+              <div>
+                &nbsp;{`(${props.zigbee2mqttDevices.length} `}
+                <Text id="integration.zigbee2mqtt.discover.device" />
```
Instead of manual concat, I think Text with field is better.
```
<Text id="integration.zigbee2mqtt.discover.device" fields={{ nbDevices: props.zigbee2mqttDevices.length }} />
```
with en.json / fr.json

integration.zigbee2mqtt.discover.device = "({nbDevices}) Devices"

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/discover-page/actions.js:
```diff
@@ -0,0 +1,130 @@
+        zigbee2mqttDevices
+      });
+    },
+    async getHouses(state) {
```
We already have common action for houses.
You can "merge" multiple actions instead of duplicate code.

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/discover-page/index.js:
```diff
@@ -0,0 +1,37 @@
+)
+class Zigbee2mqttIntegration extends Component {
+  componentWillMount() {
+    this.props.session.dispatcher.addListener(WEBSOCKET_MESSAGE_TYPES.ZIGBEE2MQTT.DISCOVER, payload => {
```
Please add both removeListener on component unmount.

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/settings-page/SettingsTab.jsx:
```diff
@@ -0,0 +1,73 @@
+                  usbPort =>
+                    usbPort.comVID && (
+                      <option value={usbPort.comPath} selected={props.zigbee2mqttDriverPath === usbPort.comPath}>
+                        {`${usbPort.comPath} - ${usbPort.comName} - ${usbPort.comVID}`}
```
`usbPort.comVID` I read COVID !! :p

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/settings-page/actions.js:
```diff
@@ -0,0 +1,96 @@
+          });
+        } else {
+          await state.httpClient.post('/api/v1/service/zigbee2mqtt/variable/ZIGBEE2MQTT_DRIVER_PATH', {
+            value: state.zigbee2mqttDriverPath
```
Can we use upper if to define `value` and avoid `post` copy?

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/setup-page/SetupTab.jsx:
```diff
@@ -0,0 +1,262 @@
+                        />
+                      )}
+                      {props.mqttRunning && props.gladysConnected && (
+                        <i style={{ color: '#0f0', fontSize: '24px' }} class="fe fe-check" />
```
I saw a style.css file, can we use CSS classes instead of explicit style attibute?

atrovato replied:
And if we use CSS classes, you will be able to use `classnames` import to have conditional classes and avoid to "big" if.

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/setup-page/SetupTab.jsx:
```diff
@@ -0,0 +1,262 @@
+                    </td>
+                  </tr>
+                  <tr>
+                    <td>{`Zigbee2mqtt`}</td>
```
Please prefer i10n files.

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/setup-page/SetupTab.jsx:
```diff
@@ -0,0 +1,262 @@
+                    </td>
+                  </tr>
+                  <tr>
+                    <td>{`MQTT`}</td>
```
Please prefer i10n files.

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/setup-page/SetupTab.jsx:
```diff
@@ -0,0 +1,262 @@
+                  <th>
+                    <Text id="systemSettings.containers" />
+                  </th>
+                  <th>{`Status`}</th>
```
Please prefer i10n files.

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/setup-page/SetupTab.jsx:
```diff
@@ -0,0 +1,262 @@
+                </thead>
+                <tbody class="text-center">
+                  <tr>
+                    <td>{`Gladys`}</td>
```
Please prefer i10n files (may exists in a common translation).

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/setup-page/SetupTab.jsx:
```diff
@@ -0,0 +1,262 @@
+                  <th>
+                    <Text id="integration.zigbee2mqtt.setup.link" />
+                  </th>
+                  <th>{`Status`}</th>
```
Please prefer i10n files.

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/setup-page/SetupTab.jsx:
```diff
@@ -0,0 +1,262 @@
+                </thead>
+                <tbody class="text-center">
+                  <tr>
+                    <td>{`Gladys - MQTT`}</td>
```
Please prefer i10n files.

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/setup-page/SetupTab.jsx:
```diff
@@ -0,0 +1,262 @@
+                        <i style={{ color: '#0f0', fontSize: '24px' }} class="fe fe-check" />
+                      )}
+                      {props.zigbee2mqttExist && !props.zigbee2mqttConnected && (
+                        <i style={{ color: '#f00', fontSize: '24px' }} class="fe fe-x" />
```
Please prefer CSS classes (and classnames import).

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/setup-page/SetupTab.jsx:
```diff
@@ -0,0 +1,262 @@
+                    </td>
+                  </tr>
+                  <tr>
+                    <td>{`MQTT - Zigbee`}</td>
```
Please prefer i10n files.

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/setup-page/SetupTab.jsx:
```diff
@@ -0,0 +1,262 @@
+                          style={{ color: '#00f', backgroundColor: '#00f', borderColor: '#00f', height: 2, width: 40 }}
+                        />
+                      )}
+                      {props.zigbee2mqttRunning && props.zigbee2mqttConnected && (
```
Please prefer CSS classes (and classnames import).

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/setup-page/SetupTab.jsx:
```diff
@@ -0,0 +1,262 @@
+                      {props.zigbee2mqttExist && props.zigbee2mqttConnected && (
+                        <i style={{ color: '#0f0', fontSize: '24px' }} class="fe fe-check" />
+                      )}
+                      {props.zigbee2mqttExist && !props.zigbee2mqttConnected && (
```
Please prefer CSS classes (and classnames import).

cicoub13 replied:
Fixed

atrovato commented on front/src/routes/integration/all/zigbee2mqtt/setup-page/actions.js:
```diff
@@ -0,0 +1,171 @@
+          container.created_at_formatted = dayjs(container.created_at * 1000)
+            .locale(state.user.language)
+            .fromNow();
+          if (container.name === '/gladys-z2m-zigbee2mqtt') {
```
I'm not sur we want this logic at front side. Maybe the /api/v1/service/zigbee2mqtt/status can compute all of this in backend?

cicoub13 replied:
Removed (it's already in status)

atrovato commented on server/services/zigbee2mqtt/api/zigbee2mqtt.controller.js:
```diff
@@ -0,0 +1,149 @@
+    } catch (e) {
+      logger.error('Error while connecting to MQTT:', e);
+      response = false;
+      gladys.event.emit(EVENTS.WEBSOCKET.SEND_ALL, {
```
Not really fan of functional stuffs in controller.

cicoub13 replied:
Removed (it's already there in installMqttContainer)

atrovato commented on server/services/zigbee2mqtt/docker/gladys-z2m-mqtt-container.json:
```diff
@@ -0,0 +1,31 @@
+{
+  "name": "gladys-z2m-mqtt",
+  "Image": "eclipse-mosquitto:latest",
```
Please fix mosquitto version, we had a problem once mosquitto did upgrade.

cicoub13 replied:
Fixed to major 2

atrovato commented on server/services/zigbee2mqtt/lib/getPermitJoin.js:
```diff
@@ -0,0 +1,13 @@
+ * init();
+ */
+function getPermitJoin() {
+  return this.z2mPermitJoin;
```
Here you don't `Prepares service and starts connection with broker`
It's only a getter.

cicoub13 replied:
Documentation updated

atrovato commented on server/services/zigbee2mqtt/lib/setValue.js:
```diff
@@ -0,0 +1,53 @@
+      zigbeeValue = value ? `{"state": "ON"}` : `{"state": "OFF"}`;
+      break;
+    case 'brightness':
+      zigbeeValue = `{"brightness": ${value}}`;
```
For information, you can `JSON.stringify` to transform objects to string.
It can be useful for "big" objects. Here it's fine.

atrovato reviewed (changes requested):
Ok, look quite good.
Some little code move and file deletion.
It will add many many many device compatibility, nice to have.

I'm sure community validates it.

cicoub13 committed:
feat(zigbee2mqtt): Take PR feedbacks

- i18n and localization fixes
- reuse common getHouse action
- removeListeners in component unmount
- use classNames and external styles
- remove unused code
- fix mosquitto version to 2
- add demo.json
- clean images and unwanted tracked files

atrovato reviewed (dismissed):
Good for me. Ensure changes don't break anything, and go for merge!!

cicoub13 committed:
feat(zigbee2mqtt): Manage SQL_FILE_PATH to store files

cicoub13 committed:
feat(zigbee2mqtt): Deactivate buttons if service not configured

cicoub13 committed:
feat(zigbee2mqtt): Add TuYa TS0011 device

LM1LC3N7 reviewed (dismissed):
Woaw it was fast!
Why did you keep the comment about the smart plug and not the other one for TS0011? You should keep or remove both.

cicoub13 committed:
feat(zigbee2mqtt): Manage click button values

cicoub13 committed:
fix(zigbee2mqtt): Fix getHouses common action usage

Signed-off-by: Cyril Beslay <cicoub13@gmail.com>

cicoub13 committed:
fix(zigbee2mqtt): Fix some Xiaomi Devices

cicoub13 committed:
docs(zigbee2mqtt): Update documentation link

Pierre-Gilles reviewed (commented):
Hi! Thanks for your video @cicoub13, this is really a high quality PR, it looks perfect to me :) 

It looks really good to me and I'm ok to merge this when this: https://community.gladysassistant.com/t/zigbee2mqtt-image-docker-de-test-basee-gladys-v4/5746/278?u=pierre-gilles

is changed in the PR :)

cicoub13 committed:
feat(docker): Add getContainerMounts in system

cicoub13 committed:
feat(zigbee2mqtt): Path management

Create files on host regarding mount configuration

Pierre-Gilles committed:
Merge branch 'master' into zigbee2mqtt-service

Pierre-Gilles reviewed (approved):
Looks good to me! 

Thanks again for the wonderful work on this PR 👏

Now the next steps are going to be: 

- Merge on master (I'm doing this right now)
- Dev build of Gladys
- Extensive test of the dev image by the community. We have lots of new things in the coming release (not just Zigbee2mqtt).

If everything is doing going, we'll be able to release that in Gladys asap! 

Thanks again for your work 👏
